core/src/main/java/org/mule/runtime/core/internal/metadata/cache/DefaultPersistentMetadataCacheManager.java

```diff
@@ -0,0 +1,148 @@
+ * This implementation provides a way of managing persistent {@link MetadataCache}s
+ * that are stored at container-level using the ObjectStore as handler for the persistence.
+ *
+ * Cache's are <b>never evicted</b>, and will be cleared only when an explicit disposal is invoked.
+ *
+ * @since 4.1.4, 4.2.0
+ */
+public class DefaultPersistentMetadataCacheManager implements MetadataCacheManager, Startable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultPersistentMetadataCacheManager.class);
+  public static final String PERSISTENT_METADATA_SERVICE_CACHE = "_mulePersistentMetadataService";
+
+  @Inject
+  @Named(OBJECT_STORE_MANAGER)
+  private ObjectStoreManager objectStoreManager;
+
+  @Inject
+  private LockFactory lockFactory;
+
+  private LazyValue<ObjectStore<MetadataCache>> metadataStore;
+
+  @Override
+  public void start() {
+    metadataStore = new LazyValue<>(() -> objectStoreManager.getOrCreateObjectStore(PERSISTENT_METADATA_SERVICE_CACHE,
+                                                                                    ObjectStoreSettings.builder()
+                                                                                        .persistent(true)
+                                                                                        .build()));
+  }
+
+  @Override
+  public MetadataCache getOrCreateCache(String id) {
+    return withKeyLock(id, key -> {
+      try {
+        if (metadataStore.get().contains(key)) {
+          return metadataStore.get().retrieve(key);
+        }
+
+        LOGGER.debug("Creating new cache " + id);
+        DefaultMetadataCache metadataCache = new DefaultMetadataCache();
+        metadataStore.get().store(key, metadataCache);
+        return metadataCache;
+
+      } catch (Exception e) {
+        String msg = String.format("An error occurred while retrieving the MetadataCache with ID '%s': %s",
+                                   id, e.getMessage());
+        LOGGER.error(msg);
+        throw new RuntimeException(msg, e);
+      }
+    });
+  }
+
+  @Override
+  public void updateCache(String id, MetadataCache cache) {
+    withKeyLock(id, key -> {
+      try {
+        LOGGER.debug("updateCache Key: " + id);
+        if (metadataStore.get().contains(key)) {
+          metadataStore.get().remove(key);
+        }
+        metadataStore.get().store(key, cache);
+      } catch (Exception e) {
+        String msg = String.format("An error occurred while updating the MetadataCache with ID '%s': %s",
+                                   id, e.getMessage());
+        LOGGER.error(msg);
+        throw new RuntimeException(msg, e);
+      }
+      return null;
+    });
+  }
+
+  @Override
+  public void dispose(String keyHash) {
+    withKeyLock(keyHash, key -> {
+      try {
+        metadataStore.get().remove(key);
+      } catch (ObjectDoesNotExistException e) {
+        LOGGER.debug("No exact match found: " + key);
```

Comment:
I would change it to something more explicit, key `More than one match for key: $key. Disposing all matching values`